zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,289 @@
+            .to_state();
+
+        let mut buf = vec![];
+
+        self.trans
+            .inputs()
+            .map(|input| match input {
+                TransparentInput::PrevOut { sequence, .. } => sequence,
+                TransparentInput::Coinbase { sequence, .. } => sequence,
+            })
+            .try_for_each(|sequence| (&mut buf).write_u32::<LittleEndian>(*sequence))?;
+
+        hash.update(&buf);
+
+        writer.write_all(hash.finalize().as_ref())
+    }
+
+    /// Writes the u256 hash of the transactions outputs to the provided Writer
+    fn hash_outputs<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+        // if (self.hash_type & SIGHASH_MASK) == SIGHASH_SINGLE
+        //     || (self.hash_type & SIGHASH_MASK) == SIGHASH_NONE
+        // {
+        //     self.outputs_hash(writer)
+        // } else if (self.hash_type & SIGHASH_MASK) == SIGHASH_SINGLE
+        // //     && transparent_input.is_some()
+        // //     && transparent_input.as_ref().unwrap().0 < tx.vout.len()
+        // {
+        //     self.single_output_hash(writer)
+        // } else {
+        //     writer.write_all(&[0; 32])
+        // }
+    }
+
+    fn outputs_hash<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+    }
+
+    fn single_output_hash<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+    }
+
+    fn hash_joinsplits<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+
+        // if !tx.joinsplits.is_empty() {
+        //     writer
+        //         .write_all(&[0; 32])
+        //         .expect("write to hasher will never fail");
+        // }
+
+        // todo!()
+    }
+
+    fn hash_lock_time<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.trans.lock_time().zcash_serialize(&mut writer)
+    }
+
+    fn hash_expiry_height<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u32::<LittleEndian>(
+            self.trans
+                .expiry_height()
+                .expect("fOverwintered is always set")
+                .0,
+        )
+    }
+
+    fn hash_hash_type<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u32::<LittleEndian>(self.hash_type)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::SigHasher;
+    use crate::{
+        serialization::ZcashDeserializeInto, transaction::Transaction, types::BlockHeight, Network,
+    };
+    use color_eyre::eyre;
+    use eyre::Result;
+    use zebra_test::vectors::ZIP143_1;
+
+    macro_rules! assert_hash_eq {
+        ($expected:literal, $hasher:expr, $f:ident) => {
+            let mut buf = vec![];
+            $hasher
+                .$f(&mut buf)
+                .expect("hashing into a vec never fails");
+            assert_eq!(hex::decode($expected.as_ref()).unwrap(), buf);
+        };
+    }
+
+    #[test]
+    fn test_vec1() -> Result<()> {
```

Comment:
Good work!

I'm glad we have test vectors for some things :-)

Reply:
yea, this one felt super satisfying once I realized that they'd broken up the test vector by component parts and the impl was already so close to being split the same way. I'm really happy with how this code / this test is shaping up.